components/raftstore/src/store/worker/split_check.rs

```diff
@@ -198,10 +197,35 @@ where
             "end_key" => log_wrappers::Value::key(&end_key),
         );
         CHECK_SPILT_COUNTER.all.inc();
-
+        let bucket_check_policy = policy;
         let mut host =
             self.coprocessor
                 .new_split_checker_host(region, &self.engine, auto_split, policy);
+        if bucket_check_policy == CheckPolicy::Approximate && host.enable_region_bucket() {
+            let bucket_keys = match host.approximate_bucket_keys(region, &self.engine) {
+                Ok(keys) => keys
+                    .into_iter()
+                    .map(|k| keys::origin_key(&k).to_vec())
+                    .collect(),
+                Err(e) => {
+                    error!(%e;
+                        "failed to get approximate bucket key";
+                        "region_id" => region_id,
+                    );
+                    vec![]
+                }
+            };
+            info!("starting approximate_bucket_keys {}", bucket_keys.len());
+            if !bucket_keys.is_empty() {
+                let mut region_buckets = Buckets::default();
+                region_buckets.set_keys(::protobuf::RepeatedField::from_vec(bucket_keys));
+                let _ = self.router.send(
+                    region.get_id(),
+                    CasualMessage::RefreshRegionBuckets { region_buckets },
```

Comment:
It's unnecessary if the region has to split by scan.

Reply:
In your implementation, you first collect bucket keys by approximate way and if the region has to be splitted by scan, bucket keys are also collected by scan, so you calculate bucket keys twice if a region has to be splitted by scan and the approximate calculation is unnecessary.